azurerm/internal/services/securitycenter/device_security_group_resource.go

```diff
@@ -0,0 +1,571 @@
+				MinThreshold:   utils.Int32(minThreshold),
+				MaxThreshold:   utils.Int32(maxThreshold),
+				IsEnabled:      utils.Bool(true),
+			})
+		case security.RuleTypeQueuePurgesNotInAllowedRange:
+			result = append(result, security.QueuePurgesNotInAllowedRange{
+				TimeWindowSize: utils.String(timeWindowSize),
+				MinThreshold:   utils.Int32(minThreshold),
+				MaxThreshold:   utils.Int32(maxThreshold),
+				IsEnabled:      utils.Bool(true),
+			})
+		case security.RuleTypeTwinUpdatesNotInAllowedRange:
+			result = append(result, security.TwinUpdatesNotInAllowedRange{
+				TimeWindowSize: utils.String(timeWindowSize),
+				MinThreshold:   utils.Int32(minThreshold),
+				MaxThreshold:   utils.Int32(maxThreshold),
+				IsEnabled:      utils.Bool(true),
+			})
+		case security.RuleTypeUnauthorizedOperationsNotInAllowedRange:
+			result = append(result, security.UnauthorizedOperationsNotInAllowedRange{
+				TimeWindowSize: utils.String(timeWindowSize),
+				MinThreshold:   utils.Int32(minThreshold),
+				MaxThreshold:   utils.Int32(maxThreshold),
+				IsEnabled:      utils.Bool(true),
+			})
+		}
+	}
+	return &result, nil
+}
+
+func flattenDeviceSecurityGroupAllowListRule(input *[]security.BasicAllowlistCustomAlertRule) []interface{} {
+	if input == nil {
+		return []interface{}{}
+	}
+	result := make([]interface{}, 0)
+	for _, v := range *input {
+		var isEnabled *bool
+		var t string
+		var values *[]string
+		switch v := v.(type) {
+		case security.ConnectionToIPNotAllowed:
+			isEnabled = v.IsEnabled
+			t = string(v.RuleType)
+			values = v.AllowlistValues
+		case security.LocalUserNotAllowed:
+			isEnabled = v.IsEnabled
+			t = string(v.RuleType)
+			values = v.AllowlistValues
+		case security.ProcessNotAllowed:
+			isEnabled = v.IsEnabled
+			t = string(v.RuleType)
+			values = v.AllowlistValues
+		default:
+			continue
+		}
+		if isEnabled == nil || !*isEnabled {
+			continue
+		}
+		result = append(result, map[string]interface{}{
+			"type":   t,
+			"values": utils.FlattenStringSlice(values),
+		})
+	}
+	return result
+}
+
+func flattenDeviceSecurityGroupTimeWindowRule(input *[]security.BasicTimeWindowCustomAlertRule) []interface{} {
+	if input == nil {
+		return []interface{}{}
+	}
+	result := make([]interface{}, 0)
+	for _, v := range *input {
+		var isEnabled *bool
+		var t string
+		var timeWindowSizePointer *string
+		var minThresholdPointer, maxThresholdPointer *int32
+		switch v := v.(type) {
+		case security.ActiveConnectionsNotInAllowedRange:
+			isEnabled = v.IsEnabled
+			t = string(v.RuleType)
+			timeWindowSizePointer = v.TimeWindowSize
+			minThresholdPointer = v.MinThreshold
+			maxThresholdPointer = v.MaxThreshold
+		case security.AmqpC2DMessagesNotInAllowedRange:
+			isEnabled = v.IsEnabled
+			t = string(v.RuleType)
+			timeWindowSizePointer = v.TimeWindowSize
+			minThresholdPointer = v.MinThreshold
+			maxThresholdPointer = v.MaxThreshold
+		case security.MqttC2DMessagesNotInAllowedRange:
```

Comment:
I think we can do something better here? there is so much repetition here... all these case statements are exactly the same?

Reply:
yes, I think the swagger is not good, which caused all structs have the same property. In fact, they should promote those common property to parent struct.

Fow now, the only way I could think of is to use reflect, but that may make the codes not easy to read and maintain.

Reply:
i mean you could have a map of all these values and then if ok := map[key], ok, then do the 5 lines of assignment, or use case fallthrough - anything but this

Reply:
it seems both way could not work
1. in golang, map could not store the struct type as value
2. fallthrough could not be used in type switch

Reply:
I'm not sure i follow? you have a type that you are testing on, you can fallthrough each statement till the last one does the assingents as every case is the same? or you can have a map of valid types:

```
	switch 1 {
	case 1:
		fallthrough
	case 2:
		fallthrough
	case 3:
		fmt.Println("case 1-3")
	default:
		fmt.Println("no match paic")
	}

	valid := map[string]bool{
		"AmqpC2DMessagesNotInAllowedRange": true,
		"AmqpC2DMessagesInAllowedRange":    true,
	}
	
	if valid["AmqpC2DMessagesInAllowedRange"] {
		fmt.Println("valid - do stuff")
	} else {
		//not valid panic?
	}
```

would not either of these work?

(https://play.golang.org/p/k5Uhdml9fU7)

Reply:
for those case, it could work.
But here is a special case, we are switch the type of a struct. In type switch, fallthrough could not be used, go build will report error. Here is a simple example: https://play.golang.org/p/kc0flfjcrXg

as for using a map, we must first know the concrete type of the struct and then we could get the field ruleType. In go, it seems the only way to do dynamic type casting is to use type switch

Reply:
*sigh* ok i see it now, security.RuleTypeMqttC2DRejectedMessagesNotInAllowedRange is an actual type not a constant assigned to a type property 😪 - so absurd it didn't sink in that was the case.

Could we open an issue on the swagger/SDK and link to it both here in the expand and flatten functions? your right i don't think there is anything we can do to make this better without fixing the swagger and i'm going to guess that will take to long if it ever gets fixed?

Reply:
yes, have created an issue to track it: https://github.com/Azure/azure-sdk-for-go/issues/14282